Mathlib/Data/List/Lemmas.lean

```diff
@@ -82,4 +83,48 @@ theorem foldl_range_eq_of_range_eq {f : α → β → α} {g : α → γ → α}
     (foldl_range_subset_of_range_subset hfg.ge a)
 #align list.foldl_range_eq_of_range_eq List.foldl_range_eq_of_range_eq
 
+
+
+/-!
+  ### MapAccumr and Foldr
+  Some lemmas relation `mapAccumr` and `foldr`
+-/
+section MapAccumr
+
+theorem mapAccumr_eq_foldr (f : α → σ → σ × β) (as : List α) (s : σ) :
+    mapAccumr f as s = List.foldr (fun a (s, bs) =>
+                                    let ⟨s_out, b⟩ := f a s
+                                    (s_out, b :: bs)
+                                  ) (s, []) as := by
```

Comment:
I think doing this pattern matching in a theorem statement generally makes proofs harder and maybe it's better to just use the projections instead. I don't have super strong opinions on this, but it's certainly something I've never see in mathlib.

Reply:
Fair point, I've changed it to projections